# RAKE

**Task 1**: to compare two lists of keywords get from the text in two ways (manually and with Rake algorithm).  
**Task 2**: to improve Rake algorithm.  
**Task 3**: check Rake algorithm on Russian texts

Text is given from: http://www.bbc.com/news/business-43032542

In [1]:
import RAKE
import operator
import io
import pandas as pd

In [3]:
stoppath = "SmartStoplist.txt"

In [4]:
#initialize RAKE by providing a path to a stopwords file
rake_object = RAKE.Rake(stoppath)

FileNotFoundError: [Errno 2] No such file or directory: 'SmartStoplist.txt'

In [29]:
#run on RAKE on a given text
sample_file = io.open("Fish_farming.txt", 'r',encoding="utf-8")
text = sample_file.read()

15 manually marked keywords (person 1):

In [49]:
keywords1 = pd.read_csv("keywords_person1.csv", encoding='utf-8')

In [53]:
keywords1['len'] = pd.Series(len(i.split()) for i in keywords1['KeyWord'])

In [54]:
keywords1

KeyWord  len
0                      tech firm CageEye;    3
1                   Norwegian fish farms;    3
2                salmon feeding patterns;    3
3              automated feeding systems;    3
4   artificial intelligence technologies;    3
5                     monitor the salmon;    3
6             computer vision algorithms;    3
7                    parasitic sea louse;    3
8                        detect sea lice;    3
9                           fish farming;    2
10                artigical intelligence;    2
11                         farmed salmon;    2
12                            fish farms;    2
13                 hydro-acoustic system;    2
14                         Mr Sovegjarto;    2
15                            audio data;    2
16                          caged salmon;    2
17                        salmon farming;    2
18                    hi-tech approaches;    2
19                       fish monitoring;    2
20                         video cameras;    2
21                           lice attach;    2
22                          lice removal;    2
23                              business;    1
24                              industry;    1
25                           automationa;    1
26                                Norway;    1
27                                salmon;    1
28                                  fish;    1
29                                 noise;    1
30                                  feed;    1
31                               hi-tech.    1

15 manually marked keywords (person 2):

In [ ]:
keywords2 = pd.read_csv("keywords_person2.csv", encoding='utf-8')

In [ ]:
keywords2['len'] = pd.Series(len(i.split()) for i in keywords2['KeyWord'])

In [ ]:
keywords2

Thus, we get % coincidence between two undependently marked keyword lists.

In [36]:
#fing keywords with max word length 3
keywords = rake_object.run(text, maxWords = 3)

In [56]:
len(keywords)

155

In [37]:
print("Keywords:")
for word in keywords:
    print(word)

Keywords:
('7m-10m krone', 9.0)
('hi-tech approaches', 9.0)
('perennially problematic visitor', 9.0)
('self-guided tool', 9.0)
('hi-tech solutions', 9.0)
('mr shang hopes', 8.5)
('salmon prices soaring', 8.1)
('salmon sloshing loudly', 8.1)
('de-loused salmon', 8.1)
('believes mr sovegjarto', 8.0)
('humans simply overseeing', 8.0)
('artificial intelligence technologies', 8.0)
('computers carefully monitor', 8.0)
('parasitic sea louse', 8.0)
('tech firm cageeye', 7.75)
('hydro-acoustic system', 7.6)
('un-gobbled feed', 7.6)
('lingalaks fish farms', 7.25)
('modern fish farms', 7.25)
('automated fish monitoring', 7.0)
('feeding fish visually', 7.0)
('artificial intelligence', 5.0)
('mr sovegjarto', 5.0)
('mr folkedal', 4.5)
('fish farms', 4.25)
('farmed salmon', 4.1)
('million salmon', 4.1)
('caged salmon', 4.1)
('salmon farming', 4.1)
('fish farming', 4.0)
('big business', 4.0)
('100 million tonnes', 4.0)
('boost production', 4.0)
('cut costs', 4.0)
('feeding frenzy', 4.0)
('improve [exp

Not really keywords due to my introspection:  

7m-10m krone;  
perennially problematic visitor;  
salmon prices soaring;  
humans simply overseeing;  
mr shang hopes;  
believes mr sovegjarto;  
million salmon;  
big business;  
100 million tonnes;  
boost production;  
feeding frenzy;  
improve [expenditure];  
pellet detector;  
wrong place;  
giving scientists;  
huge variations;  
factors influencing;  
recently raised $3;  
causing damage;  
removed manually;  
fish make;  
firm lots;  
make decisions;  
make big;  
water currents;  
water temperature;  
fires lasers;  
fatal result;  
monitor;  
humans;  
make;  
water;  
system;  
fatal;  
data;  
produces;  
year;  
producers;  
turning;  
eat;  
produce;  
installed;  
cluster;  
swim;  
knowledge;  
save;  
money;  
wasted;  
developed;  
years;  
develop;  
observes;  
number;  
pellets;  
full;  
accurate;  
institute;  
bergen;  
insights;  
day;  
adds;  
process;  
rise;  
start;  
set;  
start-;  
5m;  
makes;  
pens;  
operators;  
determining;  
explains;  
approach;  
group;  
designed;  
stingray;  
hit;  
coagulates;  
milliseconds;  
boasts;  
website;  
mirror-;  
skin;  
reflects;  
swims;  
surprise;  
wealthy;  
5%;  
｣2;  
];  

**Results**:  
    
The text is about IT-technologies in fishing.  
155 keyword expression with maximal length equal to 3 is found in text with 624 words.

Almost all manually marked keywords are found (although I'd add 'salmon feeding patterns', 'Norway', 'automated feeding systems', 'fish monitoring' and 'computer vision algorithms').

Still there is quite a lot of trash (the accuracy is nearly 0.58%: 90 keywords are really keywords): mostly because of lonely words that are specific maybe (like 'monitor') but not really reflect the sense of the text alone while including them in ngrams helps (compare 'monitor' with 'fish monitoring'). 

If we take top-15 rake keywords (still with max length 3) then the intersection with GS is only 4 expressions (0.27%):  'hi-tech approaches', 'artificial intelligence technologies', 'tech firm cageeye' and 'parasitic sea louse'.  
Although except for several totally not key expressions ('7m-10m krone' and 'perennially problematic visitors') it's still more or less about the same but not exactly. For example: it's important to include to keywords personalities that are mentioned in article ('mr shang') but there is no need to include a verb within it (compare 'mr shang' and 'mr shang hopes', 'mr sovergjarto' and 'believes mr sovegjarto'). Similarly, 'salmon' itself should be included in keywords cause the whole text is about the salmon while 'salmon sloshing loudly' is not really key expression (adverb is definutely unnecessary and sloshing is not the main topic of text as well).


In [ ]:
('7m-10m krone', 9.0)
('hi-tech approaches', 9.0)
('perennially problematic visitor', 9.0)
('self-guided tool', 9.0)
('hi-tech solutions', 9.0)
('mr shang hopes', 8.5)
('salmon prices soaring', 8.1)
('salmon sloshing loudly', 8.1)
('de-loused salmon', 8.1)
('believes mr sovegjarto', 8.0)
('humans simply overseeing', 8.0)
('artificial intelligence technologies', 8.0)
('computers carefully monitor', 8.0)
('parasitic sea louse', 8.0)
('tech firm cageeye', 7.75)

### Rake Russian

In [3]:
#initialize RAKE by providing a path to a stopwords file
rake_object_russian = RAKE.Rake("stoplist_ru.txt")

In [11]:
#run on RAKE on a given text
sample_file = io.open("технологии_рыбалка.txt", 'r')
text_ru = sample_file.read()

In [12]:
#fing keywords with max word length 3
keywords_ru = rake_object_russian.run(text_ru, maxWords = 3)

In [13]:
keywords_ru

[('первыми занятиями человека', 9.0),
 ('собирания всяких корешков', 9.0),
 ('остаются самыми консервативными', 9.0),
 ('делают орудия труда', 9.0),
 ('большей частью увлечения', 9.0),
 ('пошло развитие гаджета', 9.0),
 ('вооружение рыбаков поступили', 9.0),
 ('высокая разрешающая способность', 9.0),
 ('проводить поиск рыбы', 9.0),
 ('предмет наличия рыбы', 9.0),
 ('оповещают рыболова писком', 9.0),
 ('дорогие модели идут', 9.0),
 ('кучу рыбацких наворотов', 9.0),
 ('способны воспроизводить звуки', 9.0),
 ('издаваемые мелкой рыбкой', 9.0),
 ('электронные донные приманки', 9.0),
 ('имитация работы жабр', 8.5),
 ('предлагают беспроводные эхолоты', 8.25),
 ('эхолоты позволяют работать', 8.25),
 ('показа технологии ebs', 8.25),
 ('установить бесплатное приложение', 8.0),
 ('предметов пользования карпятников', 8.0),
 ('по-мощью рыбак', 7.666666666666667),
 ('берегу появятся рыбаки', 7.166666666666667),
 ('рыбаки карпятники', 4.5),
 ('имитация рачка', 4.5),
 ('ebs original', 4.25),
 ('ebs™ c

Ключевые биграммы и триграммы в русском тексте rake находит гораздо хуже. Похоже, что в русском тройные коллокации, в отличие от английского, в принципе менее устойчивы и реже употребляются вместе. Особенно выделяются ненужные сочетания с глаголами. Но встречаются и подходящие: навигационный прибор, крупная рыба. Однако отдельные слова все же лучше отражают смысл текста.